nbresuse/api.py

```diff
@@ -68,5 +84,36 @@ def get_cpu_percent(p):
             # about dead processes having no CPU usage
             except:
                 return 0
-
         return sum([get_cpu_percent(p) for p in all_processes])
+
+    @staticmethod
+    def __get_docker_physical_memory():
+        with open('/sys/fs/cgroup/memory/memory.limit_in_bytes') as limit:
+            mem_limit = int(limit.read().strip())
+        return mem_limit
+
+    @staticmethod
+    def __get_docker_memory_usage():
+        with open('/sys/fs/cgroup/memory/memory.usage_in_bytes') as usage:
+            all_used = int(usage.read().strip())
+        return all_used
+
+    @staticmethod
+    def __get_docker_real_rss_memory():
+        real_rss = 0
+        with open('/sys/fs/cgroup/memory/memory.stat') as stats:
+            for line in stats.readlines():
+                stat = line.split()
+                if stat[0] in ['rss', 'inactive_file', 'active_file']:
+                    real_rss = real_rss + int(stat[1])
+        return real_rss
+
+    @staticmethod
+    def __if_dockerized():
+        if os.path.exists('/.dockerenv'):
```

Comment:
Is this a reliable way of detecting whether it is running in Docker? Does this file exist with other container engines?

Reply:
It's not present on Podman.
I found this stackoverflow question which provides a bit of background: https://superuser.com/questions/1021834/what-are-dockerenv-and-dockerinit

Reply:
I just checked with Kubernetes K3S too, it's also not present.

However it looks like `/sys/fs/cgroup/memory/memory.limit_in_bytes` is set to &lt;very large number&gt; if there's no limit:
```
k3os-5708 [~]$ kubectl run tmp-shell --rm -i --tty --image busybox -- /bin/sh
If you don't see a command prompt, try pressing enter.
/ # cat /.dockerenv
cat: can't open '/.dockerenv': No such file or directory
/ # cat /sys/fs/cgroup/memory/memory.limit_in_bytes
9223372036854771712
```
```
k3os-5708 [~]$ kubectl run tmp-shell --rm -i --tty --image busybox --limits=memory=10Mi -- /bin/sh
If you don't see a command prompt, try pressing enter.
/ # cat /sys/fs/cgroup/memory/memory.limit_in_bytes
10485760
```
So one solution might be too see if `/sys/fs/cgroup/memory/memory.limit_in_bytes` exists, and if so use `min(/sys/fs/cgroup/memory/memory.limit_in_bytes, total-system-memory)`?

It's also worth noting that cgroups2 is different, I haven't figure out how to find the memory limit there.